pkg/vm/engine/tae/txn/txnbase/txnctx.go

```diff
@@ -85,6 +85,30 @@ func (ctx *TxnCtx) GetCommitTS() types.TS {
 	return ctx.CommitTS
 }
 
+func (ctx *TxnCtx) GetTxnState(waitIfcommitting bool) txnif.TxnState {
+	ctx.RLock()
```

Comment:
Break this function to two, each with a defer?

Reply:
It can be changed to two, but it is recommended not to use defer if you need wait committing. Two operations are required. The first operation is to use the read lock to check the status. If it is not committing, it will return directly. However, if it is committing, you need to release the read lock, call the lock interface in cond, and finally release cond. lock. Of course we can just use cond.L.Lock(), then defer cond.L.Unlock(). But cond.L.Lock() is exclusive.

Reply:
I would do 
func (ctx *TxtCtx) GetTxnState(wait bool) txnif.TxtState {
    if quickOK, state := quickRlock(wait); quickOK {
         return state 
     }
    return slowWlock()
}

Then defer unlock in quick/slow call.

That said, is this code really correct?   What if you read state != txnif.TxtStateCommitting, returned state, then someone change the state to Committing?

Reply:
At least here we need quite some comment to convince people this is actually right thing.

Reply:
> 
If `state != txnif.TxtStateCommitting`, the txn should be active, rollbacked, or committed. And only active txn would be changed later. 
GetTxnState(false) is used to get the txn current state atomically.
GetTxnState(true) is used during snapshot reads. If `Active` is currently returned, this value will definitely not be used, because even if it becomes `committing` later, the timestamp would be larger than the current read timestamp.
There are few comments in the code now, we will add the necessary comments in this version